File: TicTacToe/Swift/Board.swift
enum GameStatus {
    case winX
    case winO
    case tie
    case Continue
}

typealias Player = Character
typealias BoardRow = [Player]

let X: Character = "X", O: Character = "O", NO_ONE: Character = " "

let winPositions: [[Int]] = [
    [1, 2, 3], [4, 5, 6], [7, 8, 9],
    [1, 4, 7], [2, 5, 8], [3, 6, 9],
    [1, 5, 9], [3, 5, 7]
]

struct Board {
    private var firstRow, secondRow, thirdRow: BoardRow
    
    init() {
        firstRow = Array(repeating: NO_ONE, count: 3)
        secondRow = Array(repeating: NO_ONE, count: 3)
        thirdRow = Array(repeating: NO_ONE, count: 3)
    }
    
    func display() {
        print("\t\t\t\(firstRow[0]) | \(firstRow[1]) | \(firstRow[2])")
        print("\t\t\t---------")
        print("\t\t\t\(secondRow[0]) | \(secondRow[1]) | \(secondRow[2])")
        print("\t\t\t---------")
        print("\t\t\t\(thirdRow[0]) | \(thirdRow[1]) | \(thirdRow[2])")
    }
    
    private func getPlayer(atPosition position: Int) -> Player! {
        return (
            position <= 3 ? firstRow[position - 1] :
            position <= 6 ? secondRow[position - 4] :
            position <= 9 ? thirdRow[position - 7] : nil
        )
    }
    
    func isMoveLegal(atPosition position: Int) -> Bool {
        return self.getPlayer(atPosition: position) == NO_ONE
    }
    
    mutating func makeMove(atPosition position: Int, withPiece piece: Player) {
        if position <= 3 {
            firstRow[position - 1] = piece
        } else if position <= 6 {
            secondRow[position - 4] = piece
        } else {
            thirdRow[position - 7] = piece
        }
    }
    
    mutating func unmakeMove(atPosition position: Int) {
        if position <= 3 {
            firstRow[position - 1] = NO_ONE
        } else if position <= 6 {
            secondRow[position - 4] = NO_ONE
        } else {
            thirdRow[position - 7] = NO_ONE
        }
    }
    
    func checkGameStatus() -> GameStatus {
        for winPos in winPositions {
            var firstElement, secondElement, thirdElement: Player
            firstElement = self.getPlayer(atPosition: winPos[0])
            secondElement = self.getPlayer(atPosition: winPos[1])
            thirdElement = self.getPlayer(atPosition: winPos[2])
            if firstElement == secondElement && secondElement == thirdElement  &&
                thirdElement != NO_ONE {
                return (thirdElement == X ? .winX : .winO)
            }
        }
        for winPos in winPositions {
            for winp in winPos {
                if self.getPlayer(atPosition: winp) == NO_ONE {
                    return .Continue
                }
            }
        }
        return .tie
    }
}

File: TicTacToe/NodeJS/Board.js
const { gameStatus, gamePiece } = require('./enums');

const { WIN_O, WIN_X, TIE, CONTINUE } = gameStatus;
const { X, EMPTY } = gamePiece;

const winningPositions = [
    [1, 2, 3], [4, 5, 6], [7, 8, 9],
    [1, 4, 7], [2, 5, 8], [3, 6, 9],
    [1, 5, 9], [3, 5, 7]
];

class Board {
    constructor() {
        this.firstRow = [];
        this.secondRow = [];
        this.thirdRow = [];
        for (let i = 0; i < 3; i++) {
            this.firstRow[i] = EMPTY;
            this.secondRow[i] = EMPTY;
            this.thirdRow[i] = EMPTY;
        }
    }

    display() {
        console.log(`\t\t\t${this.firstRow[0]} | ${this.firstRow[1]} | ${this.firstRow[2]}`);
        console.log("\t\t\t---------");
        console.log(`\t\t\t${this.secondRow[0]} | ${this.secondRow[1]} | ${this.secondRow[2]}`);
        console.log("\t\t\t---------");
        console.log(`\t\t\t${this.thirdRow[0]} | ${this.thirdRow[1]} | ${this.thirdRow[2]}\n`);
    }

    _getPiece(position) {
        return (
            position <= 3 ? this.firstRow[position - 1] :
            position <= 6 ? this.secondRow[position - 4] :
            position <= 9 ? this.thirdRow[position - 7] : null
        );
    }

    isMoveLegal(position) {
        return this._getPiece(position) == EMPTY;
    }

    makeMove(position, piece) {
        if (position <= 3) {
            this.firstRow[position - 1] = piece;
        } else if (position <= 6) {
            this.secondRow[position - 4] = piece;
        } else {
            this.thirdRow[position - 7] = piece;
        }
    }

    unmakeMove(position) {
        if (position <= 3) {
            this.firstRow[position - 1] = EMPTY;
        } else if (position <= 6) {
            this.secondRow[position - 4] = EMPTY;
        } else {
            this.thirdRow[position - 7] = EMPTY;
        }
    }

    checkGameStatus() {
        for (const winPos of winningPositions) {
            const firstElement = this._getPiece(winPos[0]);
            const secondElement = this._getPiece(winPos[1]);
            const thirdElement = this._getPiece(winPos[2]);
            if (firstElement == secondElement && secondElement == thirdElement &&
                thirdElement != EMPTY) {
                return thirdElement == X ? WIN_X : WIN_O;
            }
        }
        for (const winPos of winningPositions)
            for (const winp of winPos)
                if (this._getPiece(winp) == EMPTY)
                    return CONTINUE;
        return TIE;
    }
}

module.exports = Board;
